clang/lib/CodeGen/CGExprConstant.cpp

```diff
@@ -870,6 +904,15 @@ bool ConstStructBuilder::Build(const APValue &Val, const RecordDecl *RD,
     if (!EltInit)
       return false;
 
+    if (ZeroInitPadding) {
+      if (!DoZeroInitPadding(Layout, FieldNo, *Field, SizeSoFar,
+                             IsFlexibleArray, AllowOverwrite))
+        return false;
+      if (IsFlexibleArray)
```

Comment:
I think I've figured out the way IsFlexibleArray works, but it's sort of awkward.  Do you really need to use a different codepath for flexible-array fields, vs. ordinary fields?  It seems like it would be simpler to just use getTypeAllocSize() in both cases.  And the name "IsFlexibleArray" is a bit confusing because isn't really reliably detecting flexible arrays; there are other zero-size fields.

Reply:
Changed IsFlexibleArray to FieldSize.
If you mean to use getTypeAllocSize() in DoZeroInitPadding(),
I don't know how to replace ASTContext::getTypeSizeInChars(clang::QualType T) with DataLayout::getTypeAllocSize(llvm::Type *Ty).

If you mean to always use CGM.getDataLayout().getTypeAllocSize(EltInit->getType()), there is a case when EltInit isn't avaiable (when emitting a DesignatedInitUpdateExpr with a nested InitListExpr).